### ISSUE TITLE
"Nicer" Errors back from H5PY

### ISSUE THREAD

**Issue by smason** [h5py/h5py#627]
I'm trying to display "nice" error messages back to users of my program but I can't get inside the errors I get back from `h5py` (similar to that noted in #493).  For example:

```
try:
    with h5.File("nonexist",'r') as fd:
        pass
except IOError as e:
    err = e
```

results in `err` containing a tuple with exactly one element and would require fragile parsing of the message if any details are needed.

After looking through `h5py/_errors.py` I can see that getting out the C-level `errno` would require parsing the `desc` from the bottom of the error stack, which is obviously fragile.  Still it could be nice to pass out the error descriptions (top and bottom of the stack) and HDF5 error IDs.

I'd quite like to have a custom error object that returns the full stack of errors back to the user.  Low level routines would just leave it at this, but higher level ones could catch and raise the same errors as are currently done.

**Comment by andrewcollette** [comment]
I agree getting just "IOError" is a bit of a pain; this is why, for example, there are so many OSError subclasses in Python 3.  However, I'm reluctant to complicate the low-level interface in this way... one of the basic design decisions for h5py is to use Python exceptions exclusively.

For generating user-facing messages, I would suggest, fragile though it may be, using e.g. a regular expression to extract some of this information from the error string.

**Comment by smason** [comment]
Thanks for getting back!  How about a set of wrapper classes that go around the errors and provide easier access to the full HDF5 error stack?

The current code gives (partial) access to the top and bottom error messages, but only after merging them both into a single string.  How about using multiple-inheritance to have HDF5 errors derived from both the currently employed Python exception as well as a h5py specific one, i.e. something [like this][mie]:

```python
class H5Exception(Exception):
    def __init__(self, desc, maj_num=None, min_num=None, func_name=None, file_name=None, line=None, parent=None, **kwds):
        super().__init__(desc, **kwds)
        self.maj_num   = maj_num
        self.min_num   = min_num
        self.func_name = func_name
        self.file_name = file_name
        self.line      = line
        self.desc      = desc
        self.parent    = parent

class H5IOError(IOError,H5Exception):
    def __init__(self, desc, **kwds):
        H5Exception.__init__(self, desc, **kwds)
        IOError.__init__(self, desc)
````

The idea being that users can catch either an `IOError` or the `h5py` variant.  Also the `parent` member gives access to the whole stack of errors from the HDF5 library which can be easily pulled out.

[mie]: https://www.reddit.com/r/Python/comments/1k9ygb/multiple_inheritance_for_library_exceptions/

**Comment by andrewcollette** [comment]
What you're describing is almost exactly how h5py worked back in the 1.3.X days; there was a base exception type (H5Error, as I recall), and we had a bunch of subclasses inherited from both that and the native Python exceptions.

The story of that approach is that we nuked it to make the code base simpler.  I never met anyone who used the custom error classes or attributes, although I agree it would help your use case.

In other words, whether to add this to h5py is less a feasibility question and more of a question of how much of the HDF5 API we want to expose, and how much complexity we want to introduce into the error system.  For those reasons I would discourage it, unless many others are strongly in favor.  You could try posting on the mailing list or pinging other developers here to see what folks think.
